src/values/ArrayValue.js

```diff
@@ -9,18 +9,88 @@
 
 /* @flow strict-local */
 
-import type { Realm } from "../realm.js";
+import type { Effects, Realm } from "../realm.js";
 import type { PropertyKeyValue, Descriptor, ObjectKind } from "../types.js";
-import { AbstractValue, ObjectValue, StringValue, NumberValue, Value } from "./index.js";
+import {
+  AbstractValue,
+  BoundFunctionValue,
+  ECMAScriptSourceFunctionValue,
+  NumberValue,
+  ObjectValue,
+  StringValue,
+  Value,
+} from "./index.js";
 import { IsAccessorDescriptor, IsPropertyKey, IsArrayIndex } from "../methods/is.js";
-import { Properties, To } from "../singletons.js";
+import { Havoc, Properties, To, Utils } from "../singletons.js";
 import { type OperationDescriptor } from "../utils/generator.js";
 import invariant from "../invariant.js";
 
+type PossibleNestedOptimizedFunctions = [
+  { func: BoundFunctionValue | ECMAScriptSourceFunctionValue, thisValue: Value },
+];
+
+function handleNestedOptimizedFunctions(
+  realm: Realm,
+  abstractArrayValue: ArrayValue,
+  possibleNestedOptimizedFunctions: PossibleNestedOptimizedFunctions
+): void {
+  for (let { func, thisValue } of possibleNestedOptimizedFunctions) {
+    let funcToModel = func;
+    if (func instanceof BoundFunctionValue) {
+      funcToModel = func.$BoundTargetFunction;
+      thisValue = func.$BoundThis;
+    }
+    invariant(funcToModel instanceof ECMAScriptSourceFunctionValue);
+    let funcCall = Utils.createModelledFunctionCall(realm, funcToModel, undefined, thisValue);
+    let hadSideEffects = false;
+    // We take the modelled function and wrap it in a pure evaluation so we can check for
+    // side-effects that occur when evaluating the function. If there are side-effects, then
+    // we don't try and optimize the nested function.
+    let pureFuncCall = () =>
+      realm.evaluatePure(funcCall, () => {
+        hadSideEffects = true;
+      });
+    let effects = realm.evaluateForEffects(pureFuncCall, null, "temporalArray nestedOptimizedFunction");
+    if (hadSideEffects) {
+      // If the nested optimized function had side-effects, then havoc the function value
+      Havoc.value(realm, func);
+    } else {
+      // Check if effects were pure then add them
+      if (abstractArrayValue.nestedOptimizedFunctionEffects === undefined) {
+        abstractArrayValue.nestedOptimizedFunctionEffects = new Map();
+      }
+      abstractArrayValue.nestedOptimizedFunctionEffects.set(funcToModel, effects);
+      realm.collectedNestedOptimizedFunctionEffects.set(funcToModel, effects);
+    }
+  }
+}
+
+function createAbstractArrayValue(
+  realm: Realm,
+  intrinsicName: string,
+  possibleNestedOptimizedFunctions?: PossibleNestedOptimizedFunctions
+): ArrayValue {
```

Comment:
The function name suggests that it returns an abstract value. It seems to do no such thing.